File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.java
import java.lang.Math;

class Solution {
    public int solution(int a, int b, int c, int d) {
        // 네 주사위가 모두 같은 경우
        if (a == b && a == c && a == d) return 1111 * a;

        // 세 주사위가 같고, 나머지 하나가 다를 경우
        if (a == b && a == c) return (int) Math.pow(10 * a + d, 2);
        if (a == b && a == d) return (int) Math.pow(10 * a + c, 2);
        if (a == c && a == d) return (int) Math.pow(10 * a + b, 2);
        if (b == c && b == d) return (int) Math.pow(10 * b + a, 2);

        // 두 주사위씩 같은 값이 나온 경우
        if (a == b && c == d) return (a + c) * Math.abs(a - c);
        if (a == c && b == d) return (a + b) * Math.abs(a - b);
        if (a == d && b == c) return (a + b) * Math.abs(a - b);

        // 두 주사위가 같은 값이 나왔고, 나머지 두 개는 서로 다른 값인 경우
        if (a == b) return c * d;
        if (a == c) return b * d;
        if (a == d) return b * c;
        if (b == c) return a * d;
        if (b == d) return a * c;
        if (c == d) return a * b;

        // 네 주사위 값이 모두 다른 경우
        return Math.min(Math.min(a, b), Math.min(c, d));
    }
}

File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.js
function solution(a, b, c, d) {
    // 주사위 숫자를 배열로 정리하고, 각 숫자의 빈도를 계산합니다.
    const dice = [a, b, c, d];
    const countMap = {};
    
    dice.forEach(num => {
        countMap[num] = (countMap[num] || 0) + 1;
    });

    // 빈도를 기준으로 숫자를 분류합니다.
    const keys = Object.keys(countMap).map(Number);
    const counts = Object.values(countMap);

    if (keys.length === 1) {
        // 네 주사위 숫자가 모두 같음
        const p = keys[0];
        return 1111 * p;
    } else if (keys.length === 2) {
        if (counts.includes(3)) {
            // 세 주사위 숫자가 같고 하나는 다름
            const p = keys[counts.indexOf(3)];
            const q = keys[counts.indexOf(1)];
            return Math.pow(10 * p + q, 2);
        } else {
            // 두 개씩 같은 숫자가 나옴
            const [p, q] = keys;
            return (p + q) * Math.abs(p - q);
        }
    } else if (keys.length === 3) {
        // 두 개는 같고 나머지 두 개는 서로 다름
        const p = keys[counts.indexOf(2)];
        const [q, r] = keys.filter(k => k !== p);
        return q * r;
    } else {
        // 네 주사위 숫자가 모두 다름
        return Math.min(...dice);
    }
}